meja/src/patmatch.ml

```diff
@@ -0,0 +1,386 @@
+open Core_kernel
+
+type pattern_case =
+  | Pcase_empty
+  | Pcase_open
+  | Pcase_type of Type0.type_expr
+  | Pcase_tuple of pattern_case list
+  | Pcase_ctor of Ident.t * pattern_case option
+  | Pcase_int of int
+  | Pcase_record of (pattern_case * Type0.type_expr) String.Map.t
+  | Pcase_or of pattern_case list
+
+let rec pprint_case fmt case =
+  let open Format in
+  match case with
+  | Pcase_empty ->
+      fprintf fmt "---EMPTY---" (*assert false*)
+  | Pcase_open | Pcase_type _ ->
+      fprintf fmt "_"
+  | Pcase_tuple cases ->
+      fprintf fmt "(@[<hv1>@,%a@,@])"
+        (pp_print_list ~pp_sep:Ast_print.comma_sep pprint_case)
+        cases
+  | Pcase_ctor (name, None) ->
+      Ident.pprint fmt name
+  | Pcase_ctor (name, Some (Pcase_tuple _ as arg)) ->
+      fprintf fmt "%a@ %a" Ident.pprint name pprint_case arg
+  | Pcase_ctor (name, Some arg) ->
+      fprintf fmt "%a@ (@[<hv1>@,%a@,@])" Ident.pprint name pprint_case arg
+  | Pcase_int i ->
+      pp_print_int fmt i
+  | Pcase_record fields ->
+      let first = ref true in
+      fprintf fmt "{@[<hv2>" ;
+      Map.iteri fields ~f:(fun ~key:name ~data:(case, _) ->
+          if !first then first := false else Ast_print.comma_sep fmt () ;
+          fprintf fmt "%s:@ @[<hv>%a@]" name pprint_case case ) ;
+      fprintf fmt "@]}"
+  | Pcase_or cases ->
+      let pp_sep fmt () = fprintf fmt "@]@ | @[<hv0>" in
+      pp_print_list ~pp_sep pprint_case fmt cases
+
+exception Empty_case
+
+let is_case_empty = function Pcase_empty -> true | _ -> false
+
+let unifies = ref (fun _ _ _ -> assert false)
+
+let rec case_of_type_decl env typ decl =
+  let open Type0 in
+  let open Type1 in
+  let snap = Snapshot.create () in
+  assert (!unifies env typ (get_mode typ.type_mode decl.tdec_ret)) ;
+  let ret =
+    match decl.Type0.tdec_desc with
+    | TAlias {type_desc= Ttuple typs; _} ->
+        Pcase_tuple (List.map typs ~f:(fun typ -> Pcase_type typ))
+    | TAbstract | TAlias _ ->
+        failwith "Could not generate a pattern case for type."
+    | TRecord _fields ->
+        let fields =
+          String.Map.empty
+          (*List.fold ~init:String.Map.empty fields ~f:(fun map field ->
```

Comment:
I left this in because it was really handy for debugging, but it's generally (IMO) a better user experience if we only talk about fields that are referred to in the patterns. For example, the unmatched cases output for the test changes from
```
{d: Some ({b: _})}
{c: false, d: Some (_)}
{a: {a: _}, d: Some (_)}
{c: false}
```
to
```
{a: _, b: _, c: _, d: Some ({a: _, b: _, c: _, d: _})}
{a: _, b: _, c: false, d: Some (_)}
{a: {a: _, b: _, c: _, d: _}, b: _, c: _, d: Some (_)}
{a: _, b: _, c: false, d: _}
```

I've added a comment to this effect in the code.